src/target/riscv/riscv.c

```diff
@@ -1589,21 +1590,100 @@ static int riscv_trigger_detect_hit_bits(struct target *target, int64_t *unique_
 		if (riscv_reg_set(target, GDB_REGNO_TSELECT, i) != ERROR_OK)
 			return ERROR_FAIL;
 
-		uint64_t tdata1;
+		uint64_t tdata1, tdata1_test_rb;
 		if (riscv_reg_get(target, &tdata1, GDB_REGNO_TDATA1) != ERROR_OK)
 			return ERROR_FAIL;
 		int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
 
 		uint64_t hit_mask = 0;
+		bool mcontrol_hit_not_supported = false;
+		bool mcontrol6_hit_not_supported = false;
 		switch (type) {
 			case CSR_TDATA1_TYPE_LEGACY:
 				/* Doesn't support hit bit. */
 				break;
 			case CSR_TDATA1_TYPE_MCONTROL:
 				hit_mask = CSR_MCONTROL_HIT;
+				/* Check if mcontrol.hit is implemented */
+				if ((tdata1 & hit_mask) == 0) {
+					uint64_t tdata1_test =
+					set_field(tdata1, CSR_MCONTROL_HIT, 1);
+					if (riscv_reg_set(target,
+					GDB_REGNO_TDATA1, tdata1_test) != ERROR_OK)
+						return ERROR_FAIL;
+					if (riscv_reg_get(target,
+					&tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+						return ERROR_FAIL;
+					int tdata1_test_hit =
+					get_field(tdata1_test, CSR_MCONTROL_HIT);
+					int tdata1_test_rb_hit =
+					get_field(tdata1_test_rb, CSR_MCONTROL_HIT);
+					if (tdata1_test_hit != tdata1_test_rb_hit)
+						mcontrol_hit_not_supported = true;
+					if (riscv_reg_set(target, GDB_REGNO_TDATA1,
```

Comment:
This is not necessary. The function already resets hit bits after the `switch`.
https://github.com/riscv-collab/riscv-openocd/blob/0097f6b46f56551d632d87f87b551c9ff2088092/src/target/riscv/riscv.c#L1707-L1709

Reply:
this function must `hit` bit is implemented, but `hit` is not implemented in HW, reset register `tdata1` lost.

Reply:
You are right. Thanks!